### PR TITLE
[GNA] Introduce 16Byte memory alignment for LNL (GNA3.6)

### PR DESCRIPTION
**Pull request by tadamowicz**
### Details:
 - Enable 16Byte memory alignment for GNA 3.6 target

### Tickets:
 - 102861

### CODE REVIEW

**Review by rjeziers — APPROVED** [review]

File: src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        if (gnadevice) {
+            gnamem.reset(new gna_memory_float(memory::GNAFloatAllocator{},
+                                              gnadevice->getMemAlignment(),
+                                              limitations::kMemoryPageSize));
+            graphCompiler.setGNAMemoryPtr(gnamem);
+            gnadevice.reset();
+        }
+    }
+
+    size_t get_memory_REGION_INPUTS_size() {
```
Comment (mkwapul) [inline src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp:63]: it's nice to have getter be const. can we do this one?

File: src/plugins/intel_gna/tests/functional/shared_tests_instances/single_layer_tests/conv_low_precision.cpp [hunk 2]
```diff
@@ -148,9 +162,9 @@ const vector<map<string, string>> configs_3_5 = {
     {{"GNA_DEVICE_MODE", "GNA_SW_EXACT"}, {"GNA_PRECISION", "I16"}, {"GNA_EXEC_TARGET", "GNA_TARGET_3_5"}},
 };
 
-const vector<map<string, string>> configs_2_0 = {
-    {{"GNA_DEVICE_MODE", "GNA_SW_EXACT"}, {"GNA_PRECISION", "I8"}, {"GNA_EXEC_TARGET", "GNA_TARGET_2_0"}},
-    {{"GNA_DEVICE_MODE", "GNA_SW_EXACT"}, {"GNA_PRECISION", "I16"}, {"GNA_EXEC_TARGET", "GNA_TARGET_2_0"}},
+const vector<map<string, string>> configs_3_6 = {
```
Comment (sirzabek) [inline src/plugins/intel_gna/tests/functional/shared_tests_instances/single_layer_tests/conv_low_precision.cpp:165]: I don't think we should add a new configuration set at this point. It won't be triggered anyway and will bring no new tests when it's finally triggered.

**Review by sirzabek — CHANGES_REQUESTED** [review]

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp [hunk 3]
```diff
@@ -78,6 +79,20 @@ inline bool IsTransposeSupported(const std::vector<size_t>& shape) {
     return min <= 8 && max % 8 == 0 && max >= 8 && max <= transposeMaxSize;
 }
 
+inline const size_t getMemoryAlignmentBytes(target::DeviceVersion target) {
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp:82]: ok

[the remark this answers is not in the document]
Comment (mkwapul) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp:82]: by the way. what's the use of const in this context?

also consider removing it from return value of getMemAlignment().

File: src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp (hunk 1 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp:63]: ok

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp [hunk 4]
```diff
@@ -78,6 +79,20 @@ inline bool IsTransposeSupported(const std::vector<size_t>& shape) {
...
+    const std::unordered_map<target::DeviceVersion, size_t> mem_alignment_map{
```
Comment (mkwapul) [inline src/plugins/intel_gna/src/backend/gna_limitations.hpp:83]: any call to this inline function creates new map. do we like it?

File: src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    HWGeneration compile_target;
+    ngraph::Shape input_shape;
+    size_t mem_input_region_size;  // Expected memory size of the input region in bytes. For this specific model and
+                                   // when the value of input_shape_H is 1, the memory input region size can be
+                                   // calculated using below formula:
+                                   // mem_input_region_size = ALIGN8(input_shape_W)*inputPrecInBytes.
+                                   // Refer to GNAGraphCompiler::AffinePrimitive for more details.
+};
+
+std::vector<MemAlignmentTestParams> all_tests_params{
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp:31]: done

File: src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp [hunk 6]
```diff
@@ -0,0 +1,134 @@
+// Copyright (C) 2018-2023 Intel Corporation
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp:1]: done

File: src/plugins/intel_gna/tests/functional/shared_tests_instances/single_layer_tests/conv_low_precision.cpp (hunk 2 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/tests/functional/shared_tests_instances/single_layer_tests/conv_low_precision.cpp:165]: removig this test

File: src/plugins/intel_gna/src/gna_device.cpp [hunk 7]
```diff
@@ -49,7 +49,7 @@ GNADeviceHelper::GNADeviceHelper(std::shared_ptr<Target> targetIn, bool isPerfor
 
     maxLayersCount_ = retrieveMaxLayersCount();
 
-    m_mem_alignment = limitations::kMemoryAlignmentBytes;
+    m_mem_alignment = limitations::getMemoryAlignmentBytes(targetIn->get_effective_compile_target());
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/src/gna_device.cpp:52]: ok

[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/src/gna_device.cpp:52]: ok

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp (hunk 4 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp]: fixed

File: src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+INSTANTIATE_TEST_SUITE_P(MemoryAlignment_FP32,
+                         GNAPluginLoadNetworkTests,
+                         ::testing::Combine(::testing::Values(ExecutionMode::SW_FP32),
+                                            ::testing::Values(HWGeneration::UNDEFINED),
+                                            ::testing::ValuesIn(param_16B_alignment_prec_fp32)),
+                         GNAPluginLoadNetworkTests::GetTestCaseName);
+
+INSTANTIATE_TEST_SUITE_P(MemoryAlignment_GNA_3_0,
```
Comment (sirzabek) [inline src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp:122]: It's a cosmetic, but why cannot we combine these two test suites into one if the only difference is the hw generation parameter?

**Review by sirzabek — APPROVED** [review]

**Review by mryzhov — APPROVED** [review]

File: src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp [hunk 9]
```diff
@@ -60,33 +52,41 @@ class GNAPluginForMemoryAlignmentTest : public GNAPlugin {
         }
     }
 
-    size_t get_memory_REGION_INPUTS_size() {
+    const size_t get_memory_REGION_INPUTS_size() const {
```
Comment (mkwapul) [inline src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp:55]: we don't need to mark return value as const

Comment (dorloff) [reply src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp:55]: @tadamowicz , could you please take this comment into account in your next PR?

**Pull request merged**